consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+            return Err(BlockError::Unknown);
+        }
+        Ok(())
+    }
+
+    fn check_tx_inputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let mut total_value = Amount::new(0);
+        for tx in transactions {
+            for input in tx.get_inputs() {
+                let tx_index =
+                    Self::get_mainchain_index_by_outpoint(&self.db_tx, input.get_outpoint())?;
+                // If there is a wrong input index then it's cause a BlockError
+                let output = Self::get_output_by_outpoint(&self.db_tx, input.get_outpoint())?;
+
+                // Check is input has already spent
+                if tx_index
+                    .get_spent_state(input.get_outpoint().get_output_index())
+                    .map_err(BlockError::from)?
+                    == OutputSpentState::Unspent
+                {
+                    return Err(BlockError::Unknown);
+                }
+                // Check overflow
+                total_value = (total_value + output.get_value()).ok_or(BlockError::Unknown)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn check_tx_outputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        for tx in transactions {
+            for _output in tx.get_outputs() {
+                // TODO: Check tx outputs to prevent the overwriting of the transaction
+            }
+        }
+        Ok(())
+    }
+
+    // Connect new block
+    fn connect_tip(
+        &mut self,
+        new_tip_block_index: &BlockIndex,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<(), BlockError> {
+        if &self.db_tx.get_best_block_id()? != new_tip_block_index.get_prev_block_id() {
+            return Err(BlockError::Unknown);
+        }
+        let block = self.get_block_from_index(new_tip_block_index)?.expect("Inconsistent DB");
+        let transactions = block.transactions();
+
+        if best_block_id.is_some() && !block.is_genesis(self.chain_config) {
+            self.check_block_fee(transactions)?;
+            self.check_tx_inputs(transactions)?;
+        }
+        self.check_tx_outputs(transactions)?;
+        self.connect_transactions(&block)?;
+
+        match &new_tip_block_index.get_prev_block_id() {
+            Some(prev_block_id) => {
+                // To connect a new block we should set-up the next_block_id field of the previous block index
+                let mut prev_block = self
+                    .db_tx
+                    .get_block_index(prev_block_id)?
+                    .expect("Can't get block index. Inconsistent DB");
+                prev_block.set_next_block_id(block.get_id());
+                self.db_tx
+                    .set_block_index(&prev_block)
+                    .expect("Can't set block index. Inconsistent DB");
+            }
+            None => {
+                if best_block_id.is_some() {
+                    panic!("Failed to read block")
+                }
```

Comment:
Could you explain what this condition does? We're adding a new block to the main chain and the previous block doesn't exist but it's only an error if the best block exists? I don't understand why the new tip would have a `None` as its previous block and what does that have to do with `best_block_id`

Reply:
I think this is just bad programming practices showing up. 

If the previous block doesn't exist, then it's the genesis (which begs the question, why not just check if it's genesis hash, which is more reliable), and if it's genesis, then best_block_id should be unset. The problem is that the consistency of the code is really bad, and things are redone again and again and checked in different places. There's certainly a better way to write this, but I can't really bet on how much better it will get overall. It should be tried though.

Reply:
this condition work when we face incorrect block that haven't `prev_block_id`, but it's not a genesis block. Seems, we can move this condition into fn `check_block` and use here `if let`. 

@altonen @TheQuantumPhysicist what do you think guys?

Reply:
Yes I think `connect_tip()` should work with validated blocks